### ISSUE TITLE
Multi-Row Insert Support In Vtgate V3

### ISSUE THREAD

**Issue by harshit-gangal** [youtube/vitess#1826]
This request is to support multi-row insert from VTGate V3
Ex.: insert into test_table(pk, shard_key, msg) values (1, 999, 'abc'), (2, 999, 'xyz')

There are 4 cases
1. Same shard key value in all rows
2. Different shard key value having same keyspace id
3. Different shard key value pointing to same shard range
4. Different shard key value pointing to different shard range

In addition to these, it has to handle auto increment keys
1. On column other than shard key
2. On shard key column
This is called out, as currently insert id returned is integer and not integer array.

**Comment by harshit-gangal** [comment]
As per the discussion with @sougou 
insert.go in vtgate/planner will be extended to take care of multi-row insert and single row insert will be treated as special case.
router.go will be changed to handle and validate the multi-row insert as currently it is doing for single row insert

**Comment by ashudeep-sharma** [comment]
The first two cases for Multi Row Insert have been implemented in this PR https://github.com/youtube/vitess/pull/1843

As per the discussion, to implement _Different shard key value pointing to same shard range_, we need to make the following changes:

- Check if Min and Max _keyspace_id_ is in the range of a single shard or not?
- Append all the _keyspace_id's_ as comma separated list eg: INSERT INTO table VALUES ((xxx), (yyy)) /* id: 1, 2 */
- Change in bin-log filtered replication to split the query and send it to two different shards based on the _keyspace_id_ ranges.

**Comment by sougou** [comment]
I'd rephrase the first action item to: Split the insert statement according to the target shards.

**Comment by ashudeep-sharma** [comment]
This Changes have been merged to master in the following Commits:
VTGate Changes: https://github.com/youtube/vitess/commit/71979f13f228e1829f6dfd1aa36f43ea50ee7d93
Filtered Replication: https://github.com/youtube/vitess/commit/1109ce0fadb7e8553491c0ed7b4775741c37b69f
